lib/reversetunnel/agent.go

```diff
@@ -302,94 +426,190 @@ func (a *Agent) proxyTransport(ch ssh.Channel, reqC <-chan *ssh.Request) {
 	wg.Wait()
 }
 
-// runHeartbeat is a blocking function which runs in a loop sending heartbeats
-// to the given SSH connection.
-//
-func (a *Agent) runHeartbeat(conn *ssh.Client) {
-	ticker := time.NewTicker(defaults.ReverseTunnelAgentHeartbeatPeriod)
+// run is the main agent loop, constantly tries to re-establish
+// the connection until stopped
+func (a *Agent) run() {
```

Comment:
Critical function, can you expand comments. How this function keeps trying to connect to the main cluster until it has discovered all proxies. Also that it has a backoff. Also that once it's connected to a proxy then it starts processing requests from it.

Reply:
:white_check_mark: